### PR TITLE
Refactor contracts and backends

### PR DESCRIPTION
**Pull request by evansd**
This PR executes some fairly major, and somewhat brutal, surgery on backends, contracts and various related concerns. I'm keen to get this stuff straightened out before we get too much further with documentation and with integrating with providers. I'm expecting, indeed looking forward to, some pushback on some of the design decisions here. But I feel reasonably confident that the general direction of travel here is the right one.

The first major change is that TableContracts are no more. Instead all the configuration they used to contain is now included in the ehrQL table declaration. For example:
```py
@construct
class patients(PatientFrame):

    date_of_birth = Series(
        datetime.date,
        description=(
            "Patient's year and month of birth, provided in format YYYY-MM-01. "
            "The day will always be the first of the month."
        ),
        constraints=[FirstOfMonthConstraint(), NotNullConstraint()],
    )
```

Only the first argument, the type, is needed by ehrQL. But all the arguments which used to be supplied to the `contracts.Column` constructor are now accepted by `Series` and can be retrieved for e.g. docs generation or backend validation.

Secondly, backends no longer declare that individual tables implement specific contracts. Instead, the backend as a whole declares that it implements one (or several) "table namespaces" (a thing we need a proper term for). These are just a Python namespace, usually a module, containing a bunch of ehrQL tables. For instance, the TPP backend declares:
```py
implements = [databuilder.tables.tpp]
```
And this means that a dataset definition written using tables imported from `databuilder.tables.tpp` can be run against this backend.

At present, the validation here is very simple: we just check that every table in the namespace is defined by the backend, and that the definition has all the specified columns.

Note that the validation does not include comparing types, which brings me on to the third major change: backends no longer specify types at all. They were, it turns out, a misleading charade because the type you'd actually get back would be whatever the type was in the database, regardless of what you'd told SQLAlchemy the type ought to be. And we already know what the type _ought_ to be from the ehrQL schema. So we can remove the duplication and simplify the backend declarations like so:
```py
class TestBackend:
    ...

    patients = MappedTable(
        source="Patient",
        columns=dict(
            patient_id="PatID",
            date_of_birth="DateOfBirth",
        ),
    )

    practice_registrations = QueryTable(
        "SELECT patient_id, date_start, date_end FROM some_table"
    )
```

To validate the types properly we need for another command, which I think we knew we wanted anyway, along the lines of:
```
databuilder validate-backend tpp --dsn mssql://192.168.1.1/db
```

That would connect to the database and check the backend against it by confirming that all the relevant tables and columns exist and that they have the right types.

There would then be scope for extending this command to validate more and more of the constraints specified in the contracts — checking that columns only contain one of a fixed number of values, or have values with a certain range and so on.

The final thing to say is that all this is going to have an impact on docs generation. We do still generate `public_docs.json` with the same fields as before, but the things they refer to aren't quite the same as the things they used to. We'll need some coordinated decision making on how we want to present things to the user, and no doubt that will end up with changes both here and in the docs repo.

### CODE REVIEW

File: databuilder/backends/base.py [hunk 1]
```diff
@@ -39,67 +39,51 @@ def get_table_expression(self, table_name, schema):
...
-        return [
-            self._make_column(name, column) for name, column in self.columns.items()
-        ]
-
-    def _make_column(self, name, column):
-        source = column.source or name
-        type_ = TYPES_BY_NAME[column.type].value
-        sql_column = sqlalchemy.Column(source, type_, key=name)
-        return sql_column
+        self.patient_join_column = source
```
Comment (benbc) [inline databuilder/backends/base.py:76]: Previously it was a bit odd that `QueryTable.learn_patient_join()` ignored its argument and triggered something that could have happened in the constructor. Now it's even more weird that this method on the parent class sets a field that is only needed by one subclass.

I'm pretty sure that there is a cleaner way of arranging what needs to happen here, but for the life of me I can't see what it is.

File: databuilder/backends/base.py [hunk 2]
```diff
@@ -39,67 +39,51 @@ def get_table_expression(self, table_name, schema):
...
-        self.columns = columns
-        self._schema = schema
-
-    def learn_patient_join(self, source):
-        if "patient_id" not in self.columns:
-            self.columns["patient_id"] = Column("integer", source)
+        self.source_table_name = source
+        self.column_map = columns
+        # Not to be confused with the schema which defines the column names and types
+        self.db_schema_name = schema
```
Comment (benbc) [inline databuilder/backends/base.py:87]: I'm so cross about this duplicated standard usage in SQL-land.

File: databuilder/backends/base.py [hunk 3]
```diff
@@ -39,67 +39,51 @@ def get_table_expression(self, table_name, schema):
...
 
-    def get_expression(self, table_name):
-        columns = self._make_columns()
-        return sqlalchemy.table(self.source, *columns, schema=self._schema)
+    def get_expression(self, table_name, schema):
+        patient_id_column = self.column_map.get("patient_id", self.patient_join_column)
```
Comment (benbc) [inline databuilder/backends/base.py:95]: Related to my comment further up. Can you remind me why it's desirable to have two ways of specifying the patient_id column? (Either explicitly in the columns or via the `patient_join_column` on the backend class.)

File: databuilder/backends/base.py [hunk 4]
```diff
@@ -74,6 +111,21 @@ def __init__(self, query, implementation_notes=None):
         self.query = query
         self.implementation_notes = implementation_notes or {}
 
+    def validate_against_table_schema(self, schema):
+        # This is a very crude form of validation: we just check that the SQL string
+        # contains each of the column names as words. But without actually executing the
+        # SQL we can't know what it returns
```
Comment (benbc) [inline databuilder/backends/base.py:117]: lol

File: databuilder/backends/databricks.py [hunk 5]
```diff
@@ -7,6 +9,7 @@ class DatabricksBackend(BaseBackend):
 
     query_engine_class = SparkQueryEngine
     patient_join_column = "patient_id"
+    implements = [databuilder.tables.databricks]
```
Comment (benbc) [inline databuilder/backends/databricks.py:12]: This means that the backend implementer has to jigger around in two different places and line things up across them. Did you consider putting the authoritative list here and dynamically generating the modules in `databuilder.tables` instead?

File: databuilder/docs/backends.py [hunk 6]
```diff
@@ -9,9 +9,12 @@ def build_backends():
...
+        # implement might not be best described as "contracts". At the very least, they
+        # will usually live under `databuilder.tables`.
+        contract_names = [namespace.__name__ for namespace in backend.implements]
+
         yield {
             "name": backend.__name__,
-            # TODO: Re-establish a listing of the tables a given backend supports once
-            # we decide exactly how this is going to work
-            "contracts": [],
+            "contracts": contract_names,
```
Comment (benbc) [inline databuilder/docs/backends.py:19]: Does this actually _break_ the docs? Have you made a plan for next steps with Steve?

**Review by benbc — COMMENTED** [review]
Lots of comments on the details, but I too think that the direction of travel here is correct. You've done a great job of seeing through all the mess and paring it down to something much cleaner.

File: tests/unit/test_query_language.py [hunk 7]
```diff
@@ -176,6 +176,19 @@ class some_table(EventFrame):
...
+    @construct
+    class some_table(PatientFrame):
+        __tablename__ = "different-name"
+        some_int = Series(int)
+        some_str = Series(str)
+
+    assert isinstance(some_table, PatientFrame)
+    assert some_table.qm_node.name == "different-name"
+    assert isinstance(some_table.some_int, IntPatientSeries)
+    assert isinstance(some_table.some_str, StrPatientSeries)
```
[the remark this answers is not in the document]
Comment (evansd) [reply tests/unit/test_query_language.py:189]: Oh, just an over-exuberant copy paste I think. Thanks for spotting.

File: tests/unit/test_query_language.py [hunk 8]
```diff
@@ -176,6 +176,19 @@ class some_table(EventFrame):
     assert isinstance(some_table.some_str, StrEventSeries)
 
 
+def test_construct_respects_custom_table_name():
+    @construct
```
[the remark this answers is not in the document]
Comment (evansd) [reply tests/unit/test_query_language.py:180]: Yes, I agree that the status quo is undesirable. My original thought was that all the logic in this decorator can be moved to `Frame.__init__` and the decorator itself just becomes:
```py
def construct(cls: type[T]) -> T:
    return cls()
```
In which case I think it would deserve its name. The fact that we need the decorator at all is pretty irksome, so I'd like it to be as simple as possible. If for any reason we don't want to go down that route then I'd be happy to rename it.

File: databuilder/backends/base.py [hunk 9]
```diff
@@ -36,7 +36,8 @@ def validate_contracts(cls):
         """
         for name, table in cls.tables.items():
             contract = table.implements
-            contract.validate_implementation(cls, name)
+            if contract is not None:
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/backends/base.py]: Interesting. I'd say it's quite idiomatic to have `is None` checks rather than truthiness for variables with the type `Optional[T]`. In fact, relying on implicit boolean conversion has some [amusing edge cases](https://lwn.net/Articles/590299/).

File: tests/unit/test_docs.py [hunk 10]
```diff
@@ -26,8 +26,9 @@ def test_generate_docs():
     output = {b["name"] for b in data["backends"]}
     assert expected <= output
 
-    names = {contract["name"] for contract in data["contracts"]}
-    assert "PatientDemographics" in names
+    # TODO: Temporarily disabled while we work on contracts
+    # names = {contract["name"] for contract in data["contracts"]}
+    # assert "PatientDemographics" in names
```
[the remark this answers is not in the document]
Comment (evansd) [reply tests/unit/test_docs.py:31]: Oh yes, good catch — thanks.

File: databuilder/docs/backends.py [hunk 11]
```diff
@@ -2,20 +2,17 @@
...
     for backend in backends:
-        # get the full name for all implemented contracts the backend implements
-        tables = [getattr(backend, name) for name in backend.tables]
-        contract_classes = [table.implements for table in tables if table.implements]
-        contract_names = [
-            "/".join([*build_hierarchy(c), c.__name__]) for c in contract_classes
-        ]
+        # TODO: We need to think through the naming here; the things which backends
+        # implement might not be best described as "contracts". At the very least, they
+        # will usually live under `databuilder.tables`.
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/docs/backends.py:14]: Next, I think, after agreeing that this is in principle mergable. And yes, I agree it's part of the same piece of work. There's a technical awkwardness here due to the fact that the docs are in a separate repo. (I'm increasingly of the view that we'd be better off having the databuilder docs in this repo and paying some cost in complexity to "embed" them somehow in the main docs, but that's another story.) And there's also complexity due to the fact that it cuts across various people's "turf".

My view (subject to modification, of course) is that the term "contracts" should be expunged entirely. We have a single thing, "descriptions of tables and their contents", with two different directions-of-fit to the world: for researchers they're descriptive; for backend implementers they're normative.

File: databuilder/contracts/contracts.py [hunk 12]
```diff
@@ -1,33 +1,27 @@
...
+    date_of_birth = Series(
+        datetime.date,
         description="Patient's year and month of birth, provided in format YYYY-MM-01.",
         constraints=[NotNullConstraint(), FirstOfMonthConstraint()],
     )
-    sex = Column(
-        type=types.Choice("female", "male", "intersex", "unknown"),
+    sex = Series(
+        str,
+        choices=["female", "male", "intersex", "unknown"],
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/contracts/contracts.py]: Yes, this is a good point. I was thinking that if we went down this road we'd need distinct types for each set of categories, and that gets complicated fast; but you're quite right that there could be value in having a `Categorical` type distinct from `str` even if that doesn't itself track the categories (that _is_ your suggestion, right?).

I don't think we can do this immediately though, because it would need changes in ehrQL along the lines of the implicit string-to-date code in order to avoid users having to write `patients.sex == Categorical("F")`. So I think in the short term leaving it as `str` but rendering it into the docs as `Categorical` is probably the least worst option.

File: Justfile [hunk 13]
```diff
@@ -105,7 +105,7 @@ docstrings: devenv
     $BIN/pydocstyle databuilder/backends/tpp.py
 
     # only enforce classes are documented for the public facing docs
-    $BIN/pydocstyle --add-ignore=D102,D103,D105,D106 databuilder/contracts/contracts.py
+    $BIN/pydocstyle --add-ignore=D102,D103,D105,D106 databuilder/contracts/wip.py
```
[the remark this answers is not in the document]
Comment (evansd) [reply Justfile:108]: Yes, me too. Particularly since we've now got, effectively, two different TPP schemas: the WIP one here, and the one I developed for the booster study. (They were originally different sorts of thing, but this PR makes them direct competitors.)

I wondered if it would be helpful to put things under a `databuilder.tables.beta` namespace while we're still working on them so it's clear to both us and users.

File: databuilder/contracts/wip.py [hunk 14]
```diff
@@ -91,39 +81,33 @@ class wip_patient_address(EventFrame):
...
     date = Series(datetime.date)
     positive_result = Series(bool)
 
 
 @construct
-class wip_simple_patient_demographics(EventFrame):
+class simple_patient_demographics(EventFrame):
     """TODO."""
 
     __tablename__ = "patients"
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/contracts/wip.py:113]: Yes, I don't see why not.

File: databuilder/tables/__init__.py [hunk 15]
```diff
@@ -1,3 +1,7 @@
+# TODO: This should probably be removed altogether but it's being used by the test
+# study, and possibly by other example code others have written so I don't now is quite
+# the right time to remove it.
+# [1]: https://github.com/opensafely/test-age-distribution
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/tables/__init__.py:4]: Yes, absolutely. I didn't want to start creating other tickets until we had some kind of consensus that this PR was acceptable.

File: databuilder/backends/base.py (hunk 1 above)
Comment (evansd) [reply databuilder/backends/base.py:76]: Yeah, it's all a bit round the houses isn't it? I didn't want to spend too much time reworking this stuff here, beyond what was needed to get the API surface right. There may be something we could do with descriptors here to make it all drop out more cleanly.

File: databuilder/backends/base.py (hunk 2 above)
Comment (evansd) [reply databuilder/backends/base.py:87]: I don't see the problem. It's either the name for a collection of tables that's sort of like a database but not a database. Or it's a list of those tables. Or it's a list of the _columns_ of those tables. Or it's either of those things, but in ehrQL rather than SQL.

File: databuilder/backends/base.py (hunk 3 above)
Comment (evansd) [reply databuilder/backends/base.py:95]: Well, we need the facility to specify them per-table because they aren't always consistent within a backend e.g. if a backend provider imports a third-party dataset without changing the column names — as we've seen before. But then other backend providers are careful to always name them consistently and so it's a bit annoying in that case to have to specify them over again for each table.

So we've ended up having a backend-wide default which can be overridden per-table if necessary. This doesn't seem totally unreasonable to me, though I agree it's a bit inelegant.

File: databuilder/backends/base.py (hunk 4 above)
Comment (evansd) [reply databuilder/backends/base.py:117]: I originally thought as this as joke solution, and then I realised it might actually be quite helpful in warning us immediately if we add a new column without updating the SQL. I'd be happy to remove it once we have proper validation against the database if you think it's a little too whimsical.

File: databuilder/backends/databricks.py (hunk 5 above)
Comment (evansd) [reply databuilder/backends/databricks.py:12]: Well the short answer is that we can't dynamically generate the stuff in `databuilder.tables` if we want autocomplete to work. But I'm also hopeful (perhaps naively so) that there won't be too much juggling across two places. If the implementer is writing a new set of ehrQL tables to reflect what's in their database then they may not need anything beyond the `DefaultBackend` which just maps the ehrQL schema directly to the database one. And if the implementer is trying to map an existing schema to their database then most of the work will be on the backend class, with the ehrQL schema as a fixed target.

Of course, I expect the reality to be a bit more messy than that. But I'm hopeful that with a bit of tooling and some extra options for backend classes we can make life sufficiently easy for backend implementers.

File: databuilder/docs/backends.py (hunk 6 above)
Comment (evansd) [reply databuilder/docs/backends.py:19]: Possibly, and no :) But as above, I didn't want to start conversations with Steve until I had more confidence in these changes.

File: tests/unit/test_query_language.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (benbc) [reply tests/unit/test_query_language.py:180]: As long as this is a live question in your mind I'm happy to leave it to you to improve as and when you see fit

File: databuilder/backends/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/backends/base.py]: That's very funny. As I say, I don't like the auto-booleanizing for just this reason, so I'm happy to adjust my idioms.

File: databuilder/docs/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/docs/backends.py:14]: Great.

> I'm increasingly of the view that we'd be better off having the databuilder docs in this repo

Strong agree. Although I don't remember the concrete reason that led to putting them in the docs repo originally.

> it cuts across various people's "turf"

I don't think that should be a problem. It's reasonable to have different people as authorities on different bits of the same codebase.

> My view ... is that the term "contracts" should be expunged entirely.

Yes, I think that's correct.


> two different directions-of-fit

Another way to think of this is as different projections of a single concept.

File: databuilder/contracts/contracts.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/contracts/contracts.py]: > if we went down this road we'd need distinct types for each set of categories

That actually hadn't occurred to me, but under exactly the current scheme you're quite right, of course.

We're using `type` for a bunch of different things here:

* constrain the types of objects that can be use in ehrQL
* constrain the values that can be used in ehrQL (for example for dates-as-strings)
* specify conversions in translating to SQL
* (presumably, later) validate types actually found in the database

And of course there's a lot of overlap with `constraints`.

The more I think about it the more tricky this seems to be. I'm pretty much convinced that the current model is not correct, but it needs more thinking to suggest a better alternative.

I'm inclined to suggest that there should be a single object which describes all constraints (types and values) on a given column -- or maybe that it should be split into two, one for ehrQL and one for the database.

> I think in the short term leaving it as str but rendering it into the docs as Categorical is probably the least worst option.

I agree, and I think this would be good enough for now with the addition (I think) that the `choices` argument should be folded into `constraints`.

File: Justfile (hunk 13 above)
[the remark this answers is not in the document]
Comment (benbc) [reply Justfile:108]: > I wondered if it would be helpful to put things under a databuilder.tables.beta namespace

Yes, maybe it would.

I can't find the issue called something like "support future backwards incompatible changes" which envisaged some kind of versioning scheme. Do you know where that's gone, @inglesp?

Irrespective of namespacing, though, I think we should be brave in unifying what we've got asap, even if it breaks something.

File: databuilder/tables/__init__.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/tables/__init__.py:4]: :+1:

File: databuilder/backends/base.py (hunk 2 above)
Comment (benbc) [reply databuilder/backends/base.py:87]: :grinning:

File: databuilder/backends/base.py (hunk 4 above)
Comment (benbc) [reply databuilder/backends/base.py:117]: No, I think it's gives some value and it's hard to see how it could be wrong.

File: databuilder/backends/databricks.py (hunk 5 above)
Comment (benbc) [reply databuilder/backends/databricks.py:12]: Right.

I'd certainly prefer to make life easier for dataset definition authors than for backend implementers.

File: databuilder/docs/backends.py (hunk 6 above)
Comment (benbc) [reply databuilder/docs/backends.py:19]: :+1:

File: databuilder/docs/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (inglesp) [reply databuilder/docs/backends.py:14]: > the term "contracts" should be expunged entirely

I feel like the term is useful for communicating an intent: "If you can expose data that conforms to these contracts, then you can be an OpenSAFELY backend".

File: tests/unit/test_query_language.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (inglesp) [reply tests/unit/test_query_language.py:180]: +1 for `@table`

File: databuilder/docs/backends.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (benbc) [reply databuilder/docs/backends.py:14]: Perhaps there is space for the term "contract", but the current use as a parallel hierarchy to "table" seems overly complex. Could we use "contract" to refer the full set of table definitions? Or something along the lines of "the contract for this table is..."?

**Comment by evansd** [comment]
@benbc @inglesp I've just renamed the `@construct` decorator to `@table` which is obviously far more comprehensible. I only avoided it initially because I somehow had the impression that we wanted to avoid "table/column" terminology. But given that's not the case, should we change the `Series` descriptor to `Column` as well?

To make autocomplete work these table definition classes are still going to have to subclass `PatientFrame` and `EventFrame`, but if we really wanted to we could alias those to other names.

**Comment by benbc** [comment]
> should we change the `Series` descriptor to `Column` as well?

I don't think that they're synonymous, are they? `Series` captures the cleverness around patients as well.

**Comment by evansd** [comment]
But by the same logic neither are "table" and "frame".

**Comment by evansd** [comment]
On second thoughts, I probably shouldn't have raised the question at this point — it was just something that floated by while I was renaming the decorator.

I'm really keen that we have _some_ consistent story and nomenclature to present to users. But I think that's best achieved by stepping back and thinking about all the parts in aggregate as a discrete task, rather than fiddling about with names in passing as we do other things.

**Comment by evansd** [comment]
@benbc @inglesp I think this is now ready for a re-review. I'm afraid I've been liberally rebasing things so there isn't an easy way to review exactly what's changed since you last reviewed. (The [diff](https://github.com/opensafely-core/databuilder/compare/c68864d950867744da85160b54ada53433803ac8..8e116c153f59eb5c4cb86eb7c216cedbe7f7cfa1) is not particularly readable.)

Along with addressing all the smaller comments, the main changes I've made have been:

 * Removed the `choices` argument and made it a constraint
 * Renamed the constraint from `ChoiceConstraint` to `CategoricalConstraint` to make it less Django-y and more stats-y
 * Moved the WIP tables to `databuilder.tables.beta`
 * Added a `schemas` key to the documentation metadata which documents all the modules under `databuilder.tables`
 * Renamed `@construct` to `@table`

**Review by benbc — APPROVED** [review]
LGTM

File: databuilder/query_language.py [hunk 16]
```diff
@@ -476,7 +476,7 @@ def construct(cls):
             "Schema class must subclass either `PatientFrame` or `EventFrame`"
         )
 
-    table_name = cls.__name__
+    table_name = getattr(cls, "__tablename__", cls.__name__)
```
[the remark this answers is not in the document]
Comment (evansd) [reply databuilder/query_language.py:479]: Probably not, no. I can remove it. It's trivial enough to add back in if we do find a use case.

**Comment by StevenMaude** [comment]
As I mentioned in opensafely/documentation#937, this is fine to merge from a documentation perspective.

We'll have to further update things there, but:

* the changes in this PR don't break the documentation build (the contracts reference is made less helpful, but the docs do continue to build)
* none of the affected documentation is "officially" public; affected pages have a prominent warnings that the page is still being worked on
